src/psyclone/tests/domain/lfric/lfric_builtins_test.py

```diff
@@ -4172,599 +1634,287 @@ def test_X_times_Y_deduce_space(dist_mem):
+    assert len(metadata.meta_args) == 2
+    assert metadata.meta_args[0].access == "gh_write"
+    assert metadata.meta_args[0].datatype == "gh_real"
+    assert metadata.meta_args[0].function_space == "any_space_1"
+    assert metadata.meta_args[1].access == "gh_read"
+    assert metadata.meta_args[1].datatype == "gh_integer"
+    assert metadata.meta_args[1].function_space == "any_space_1"
+
+    kern = builtin_from_file("15.28.2_int_to_real_X_builtin.f90")
+    assert str(kern) == ("Built-in: int_to_real_X (convert an integer-valued "
+                         "to a real-valued field)")
+
+    # Test the 'lower_to_language_level()' method
+    code = fortran_writer(kern)
+    assert (
+        "! Built-in: int_to_real_X (convert an integer-valued to a "
+        "real-valued field)\n"
+        "f2_data(df) = REAL(f1_data(df), kind=r_def)\n") in code
+
+
+@pytest.mark.parametrize("kind_name", ["r_solver", "r_tran", "r_bl", "r_phys"])
+def test_int_to_real_x_precision(tmpdir, kind_name):
+    '''
+    Test that the built-in picks up and creates correct code for field
+    data with precision that is not the default, i.e. not 'r_def'. Try with
+    all supported LFRic real-valued field precisions to make sure all work.
+
+    '''
+    # Note that monkeypatching required to change field type, field proxy
+    # type and the field data precisions is extensive and complicated.
+    # Modifying the test algorithm is easier and more effective.
+    with open(os.path.join(BASE_PATH, "15.28.2_int_to_real_X_builtin.f90"),
+              "r", encoding='utf-8') as alg_file:
+        alg_code = alg_file.read()
+
+    # Modify the 'real'-valued field type and precision, and store the
+    # modified temporary algorithm
+    pattern = re.compile(r"\bfield_")
+    alg_code = re.sub(pattern, f"{kind_name}_field_", alg_code)
+    os.mkdir(str(tmpdir.join("tmp")))
+    tmp_fname = str(tmpdir.join("tmp", f"real_{kind_name}_X_builtin_alg.f90"))
+    with open(tmp_fname, "w", encoding='utf-8') as tmp_file:
+        tmp_file.write(alg_code)
+    tmp_file.close()
+
+    # Read and parse the modified algorithm
+    with open(tmp_fname, "r", encoding='utf-8') as alg_file:
+        _, invoke_info = parse(alg_file, api=API)
+    psy = PSyFactory(API).create(invoke_info)
     code = str(psy.gen)
 
+    # Test code generation
+    assert f"USE constants_mod, ONLY: {kind_name}, i_def" in code
+    assert (f"USE {kind_name}_field_mod, ONLY: {kind_name}_field_type, "
+            f"{kind_name}_field_proxy_type") in code
+    assert f"TYPE({kind_name}_field_type), intent(in) :: f2" in code
+    assert (f"REAL(KIND={kind_name}), pointer, dimension(:) :: "
+            "f2_data => null()") in code
+    assert f"TYPE({kind_name}_field_proxy_type) f2_proxy" in code
+    assert f"f2_data(df) = REAL(f1_data(df), kind={kind_name})" in code
+
+    # Test compilation of generated code
     assert LFRicBuild(tmpdir).code_compiles(psy)
```

Comment:
Aha! Please add this to all tests that don't already have it.

Reply:
In this case I left it because we check that the declarations are valid and matching the values from the infrastructure, which the compilation test effectively does. I now added the same in a couple other similar tests. (but even here I would argue this not really the responsibility of builtins)